src/fiber/gossip.rs

```diff
@@ -0,0 +1,2738 @@
+                    Ok(message) => {
+                        // Normally we will need to check if the message has any dependencies that are not saved yet.
+                        // But this is not necessary for the messages that we are proactively broadcasting.
+                        // These are our own channel announcements, channel updates and node announcements.
+                        let _ = state
+                            .gossip_actor
+                            .send_message(GossipActorMessage::BroadcastMessageImmediately(message));
+                    }
+                }
+            }
+
+            ExtendedGossipMessageStoreMessage::Tick => {
+                trace!(
+                    "Gossip store maintenance ticked: last_cursor = {:?} #subscriptions = {},  #messages_to_be_saved = {}",
+                    state.last_cursor,
+                    state.output_ports.len(),
+                    state.messages_to_be_saved.len(),
+                );
+
+                // These are the messages that have complete dependencies and can be sent to the subscribers.
+                let complete_messages = state.prune_messages_to_be_saved().await;
+                for (id, subscription) in state.output_ports.iter() {
+                    let messages_to_send = match subscription.filter {
+                        Some(ref filter) => complete_messages
+                            .iter()
+                            .filter(|m| &m.cursor() > filter)
+                            .cloned()
+                            .collect::<Vec<_>>(),
+                        None => complete_messages.clone(),
+                    };
+                    trace!(
+                        "ExtendedGossipMessageActor sending complete messages to subscriber #{}: number of messages = {}",
+                        id, messages_to_send.len()
+                    );
+                    for chunk in messages_to_send.chunks(MAX_NUM_OF_BROADCAST_MESSAGES as usize) {
+                        subscription
+                            .output_port
+                            .send(GossipMessageUpdates::new(chunk.to_vec()));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub enum ExtendedGossipMessageStoreMessage {
+    // A new subscription for gossip message updates. We will send a batch of messages to the subscriber
+    // via the returned output port.
+    NewSubscription(
+        Option<Cursor>,
+        RpcReplyPort<(
+            u64,
+            oneshot::Sender<()>, // A channel to notify the subscriber that the subscription is ready.
+            Arc<OutputPort<GossipMessageUpdates>>,
+        )>,
+    ),
+    // Update the subscription with a new cursor. If the outer Option is None, the subscription will be cancelled.
+    // If the inner Option is None, the subscription will start from the very latest message in the store.
+    UpdateSubscription(u64, Option<Option<Cursor>>, RpcReplyPort<()>),
```

Comment:
when will we call UpdateSubscription with the second parameter value `Some(None)`?

Reply:
It is now mostly used in the test, in which we don't want to specify a cursor, instead we ask the actor to disseminate whatever he have. Admittedly, this is not very useful. I can remove that.

Reply:
which test? I grep the test code, cannot find it.

Reply:
and, a different message `CancelSubscription` may be better

Reply:
> which test? I grep the test code, cannot find it.

See the commit https://github.com/nervosnetwork/fiber/pull/308/commits/131bfd1a527438feebe4817fc51a0e32299f2c8b here for the test code which uses `Some(None)` under the hood to receive updates for whatever comes to the store.

> and, a different message CancelSubscription may be better

Downstream consumers should use the `unsubscribe` function in https://github.com/contrun/fiber/blob/8525e016231e7649052445f8a19d6a8356fed493/src/fiber/gossip.rs#L286-L289 to unsubscribe store updates. The command `UpdateSubscription(u64, Option<Cursor>, RpcReplyPort<()>)` is only intended to be used for the implementer internally.